src/libsyntax/ext/deriving/generic/mod.rs

```diff
@@ -558,6 +652,21 @@ impl<'a> TraitDef<'a> {
                        enum_def: &EnumDef,
                        type_ident: Ident,
                        generics: &Generics) -> P<ast::Item> {
+        let mut field_tys = Vec::new();
+
+        for variant in enum_def.variants.iter() {
```

Comment:
@tamird: Unfortunately it cannot because those inner iterators are actually two different types.